src/main/java/redis/clients/jedis/DefaultJedisSocketFactory.java

```diff
@@ -1,80 +1,122 @@
 package redis.clients.jedis;
 
-import redis.clients.jedis.exceptions.JedisConnectionException;
-
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.Socket;
 import javax.net.ssl.HostnameVerifier;
 import javax.net.ssl.SSLParameters;
 import javax.net.ssl.SSLSocket;
 import javax.net.ssl.SSLSocketFactory;
-import java.io.IOException;
-import java.net.InetSocketAddress;
-import java.net.Socket;
+
+import redis.clients.jedis.exceptions.JedisConnectionException;
+import redis.clients.jedis.util.IOUtils;
 
 public class DefaultJedisSocketFactory implements JedisSocketFactory {
 
-  private String host;
-  private int port;
-  private int connectionTimeout;
-  private int soTimeout;
-  private boolean ssl;
-  private SSLSocketFactory sslSocketFactory;
-  private SSLParameters sslParameters;
-  private HostnameVerifier hostnameVerifier;
+  private String host = Protocol.DEFAULT_HOST;
+  private int port = Protocol.DEFAULT_PORT;
+  private int connectionTimeout = Protocol.DEFAULT_TIMEOUT;
+  private int soTimeout = Protocol.DEFAULT_TIMEOUT;
+  private boolean ssl = false;
+  private SSLSocketFactory sslSocketFactory = null;
+  private SSLParameters sslParameters = null;
+  private HostnameVerifier hostnameVerifier = null;
+  private HostAndPortMapper hostAndPortMapper = null;
+
+  public DefaultJedisSocketFactory() {
+  }
 
+  @Deprecated
   public DefaultJedisSocketFactory(String host, int port, int connectionTimeout, int soTimeout,
       boolean ssl, SSLSocketFactory sslSocketFactory, SSLParameters sslParameters,
       HostnameVerifier hostnameVerifier) {
-    this.host = host;
-    this.port = port;
-    this.connectionTimeout = connectionTimeout;
-    this.soTimeout = soTimeout;
-    this.ssl = ssl;
-    this.sslSocketFactory = sslSocketFactory;
-    this.sslParameters = sslParameters;
-    this.hostnameVerifier = hostnameVerifier;
+    setHost(host);
+    setPort(port);
+    setConnectionTimeout(connectionTimeout);
+    setSoTimeout(soTimeout);
+    setSsl(ssl);
+    setSslSocketFactory(sslSocketFactory);
+    setSslParameters(sslParameters);
+    setHostnameVerifier(hostnameVerifier);
+  }
+
+  public DefaultJedisSocketFactory(HostAndPort hostAndPort, JedisClientConfig config) {
+    setHostAndPort(hostAndPort);
+    if (config != null) {
+      setConnectionTimeout(config.getConnectionTimeout());
+      setSoTimeout(config.getSoTimeout());
+      setSsl(config.isSsl());
+      setSslSocketFactory(config.getSslSocketFactory());
+      setSslParameters(config.getSslParameters());
+      setHostnameVerifier(config.getHostnameVerifier());
+      setHostAndPortMapper(config.getHostAndPortMapper());
```

Comment:
It's a bit unconventional to use setters in constructors, any advantage of doing that?

Reply:
Not exactly. I saw this style is being used in commons-pool2. So I put it here at least to have some discussions :)

Reply:
IMO, still unconventional commons-pool2 doesn't make it okay :P